text/3816-bidirectional-hash-maps.md

```diff
@@ -0,0 +1,115 @@
+```rust
+use std::collections::BijectionMap;
+use std::sync::LazyLock;
+
+pub enum MyEnum {
+    VariantA,
+    VariantB,
+    VariantC,
+    VariantD,
+}
+
+const STRING_CONV_MYENUM_MAP: LazyLock<BijectionMap<&str, MyEnum>> = LazyLock::new(BijectionMap::from([("a", MyEnum::VariantA), ("b", MyEnum::VariantB), ("c", MyEnum::VariantC), ("d", MyEnum::VariantD)] /* and so on... */));
+```
+and use it here:
+
+```rust
+let from_string = STRING_CONV_MYENUM_MAP.get_value("a");
+let to_string = STRING_CONV_MYENUM_MAP.get_key(MyEnum::VariantA);
+```
+
+# Guide-Level Explanation
+A BijectionMap is a Rust collection that maps keys and values together. Unlike a HashMap, you can look up a BijectionMap both ways, by key or by value.
+
+Say you're developing a program and you need to check usernames and match them up with user emails and vice versa. You can, obviously, use two hashmaps and develop a reverser function. But why do two if you can use just one map?
+
+```rust
+let USER_BIDI_MAP = BijectionMap::<&str, &str>::new();
+for user in users {
+    USER_BIDI_MAP.insert(user.username, user.email);
+}
+let random_user_who_logged_in = "iamaverysmartperson";
+let their_email = USER_BIDI_MAP.get(random_user_who_logged_in);
+```
+
+It can also be used to convert enums to other values:
+
+```rust
+pub enum IpResponse<Value, Error> {
+   Success(Value),
+   Info(Value),
+   Err(Error)
+}
+
+// let statuscode_equals_ipresponse = blah blah blah
+// This can help us get a status code, convert it to an enum, use it in functions, and send it to a user downstream for example
+```
+# Reference-Level Explanation
+BijectionMap<K, V, S> will have both Keys and Values implement Eq and Hash, but the API is quite similar to HashMap:
+
+```rust
+let ikea_bijection_map = BijectionMap::new::<&str, u32>();
+ikea_bijection_map.insert("SVALLJARD", 8321789228129);
+ikea_bijection_map.insert("STENKOL", 182317198732918);
+ikea_bijection_map.insert("VALLHEIOR", 92131831298718);
+ikea_bijection_map.insert("HAREDROR", 92311273981273);
+
+// this user wants WELTAEK. Now which one is that?
+user.get_key("samanheik288").cart.push_back(ikea_bijection_map.get_key("WELTAEK"));
+
+// one of the delivery agents gave us this product number and this username. What does it match to?
+user.get_key("heldros8492").cart.push_back(ikea_bijection_map.get_value(923179812371));
+
+// ATTENTION! SAMANKEIV is no longer being sold!
+ikea_bijection_map.remove_key("SAMANKEIV");
+
+// We ran out of 9328193103930 and IKEA will stop selling it here. What do we do?
+ikea_bijection_map.remove_value(9328193103930);
+
+// ERWAK is a great product... but is the name being used?
+ikea_bijection_map.entry_key("ERWAK").or_insert(21389181329871);
+```
+
+# Drawbacks
+None?
+
+# Rationale and Alternatives
+An alternative is to implement two or three hashmaps for looking up each value. The BijectionMap reduces memory size and typing, and is easier to understand in terms of being explicit.
+
+# Prior art
```

Comment:
There's also the bimap crate. I'm not sure if it has a similar API or not to this one, because i haven't ever used this one before.